### PR TITLE
opengl performance fixes and improvements

### PR DESCRIPTION
**Pull request by kd-11**
This set changes the way we upload data to buffers that are updated per draw call. Buffers are replaced with large buffers that serve as pools which can be mapped without causing a synchronization stall and orphaned once they are filled up instead of using fences to sync.

Performance gains vary from game to game, but I'm getting a ~25% fps boost in most games.

### CODE REVIEW

**Comment by raven02** [comment]
@kd-11 , not too sure if it is my own issue. I got black screen almost every games.

**Comment by SakataGintokiYT** [comment]
@kd-11 Same here

File: rpcs3/Emu/RSX/GL/gl_helpers.h [hunk 1]
```diff
@@ -569,6 +580,105 @@ namespace gl
...
+		{
+			glUnmapBuffer((GLenum)m_target);
+			m_mapped_block_size = 0;
+			m_mapped_base = 0;
+		}
+
+		void reserve_and_map(u32 max_size)
+		{
+			max_size = (max_size + 4095) & ~4095;
+			auto& mapping = alloc_and_map(max_size);
```
Comment (danilaml) [inline rpcs3/Emu/RSX/GL/gl_helpers.h]: Should be `const auto&`, I believe.

**Comment by kd-11** [comment]
@SakataGintokiYT Probably doesn't like that my vao has no vertex array. I'll fix it

**Comment by kd-11** [comment]
Its an alignment issue. Uploading a fix

**Comment by kd-11** [comment]
@SakataGintokiYT @raven02 Please retest. Works now on my old nv laptop

**Comment by danilaml** [comment]
@kd-11 on AMD I get:
```
{rsx::thread} class std::runtime_error thrown: Precondition failed: (offset & 15) == 0
(in file c:\rpcs3\rpcs3\emu\rsx\gl\gl_helpers.h:673)
```
in disgaea 3 after intro.

**Comment by kd-11** [comment]
@danilaml AMD supports 4-byte buffer alignment vs nvidia's 256. I made a small error there by expecting a min alignment value of 16 hence the exception. Should be fixed now

**Comment by danilaml** [comment]
@kd now d3 works like before. The fps is about the same though (it looks like it's bottlenecked somewhere else).

**Comment by SakataGintokiYT** [comment]
@kd-11 
Now works

**Comment by kd-11** [comment]
@danilaml the only remaining opengl bottleneck is draw call count. Even then, you should have a ~5-10 fps gain although i don't own d3. Maybe your cpu is powerful enough to offset the gains anyway.

I've tested with Hitman, Afterburner climax, Prince of Persia, Outrun online, and TNT racers. Those show gains, but basic 2D games just don't show much of a difference although minimum frame rates improve a bit.

**Comment by danilaml** [comment]
@kd-11 I think d3 has some severe threadening issues (d4 is better, after pause/unpause). It never reached above 4/5 fps in opening scene for me in any render.

**Comment by kd-11** [comment]
@SakataGintokiYT @raven02 any difference on nvidia? I know nv uses a separate opengl thread (__GL_THREADED_OPTIMIZATIONS) so the effect may not be as pronounced on NV drivers as it is on AMD.

**Comment by SakataGintokiYT** [comment]
@kd-11  I think that is the same fps ;/

**Comment by kd-11** [comment]
@SakataGintokiYT Significant boost on AMD, although it is only relevant on 3D games. Anyway, checks one item off the list of graphics tasks.

**Comment by SakataGintokiYT** [comment]
@kd-11 
*Koihime Enbu / GL
Fotoszok.pl

**Comment by kd-11** [comment]
@SakataGintokiYT Test 3d games that aren't very cpu limited e.g G-Force, Hitman2. I dindnt expect much of a boost on nvidia anyway due to nvidia's multithreaded driver. For me some games have almost double the performance on an R9 270x

**Comment by rcaridade145** [comment]
I see you are using glMapBufferRange. On Nvidia it is better to use glBufferSubData.

https://git.reviewboard.kde.org/r/111354/

**Comment by kd-11** [comment]
@rcaridade145 I map with UNSYNCHRONIZED bit set to prevent the pipeline stalls. Also, from reports here, Nvidia performance is not affected negatively so I feel that it is a balanced solution instead of writing a separate path for amd, nvidia and intel

**Comment by rcaridade145** [comment]
@kd-11 Agree with balance solution point :) .

**Pull request merged**